haystack/preview/components/embedders/openai_text_embedder.py

```diff
@@ -0,0 +1,90 @@
+from typing import List, Optional, Dict, Any
+
+import openai
+
+from haystack.preview import component, default_to_dict, default_from_dict
+
+
+API_BASE_URL = "https://api.openai.com/v1"
+
+
+@component
+class OpenAITextEmbedder:
+    """
+    A component for embedding strings using OpenAI models.
+    """
+
+    def __init__(
+        self,
+        api_key: str,
+        model_name: str = "text-embedding-ada-002",
+        api_base_url: str = API_BASE_URL,
+        organization: Optional[str] = None,
+        prefix: str = "",
+        suffix: str = "",
+    ):
+        """
+        Create a OpenAITextEmbedder component.
+
+        :param api_key: The OpenAI API key.
+        :param model_name: The name of the model to use.
+        :param api_base_url: The OpenAI API Base url, defaults to `https://api.openai.com/v1`.
+        :param organization: The OpenAI-Organization ID, defaults to `None`. For more details, see OpenAI
+        [documentation](https://platform.openai.com/docs/api-reference/requesting-organization).
+        :param prefix: A string to add to the beginning of each text.
+        :param suffix: A string to add to the end of each text.
+        """
+
+        self.api_key = api_key
+        self.model_name = model_name
+        self.api_base_url = api_base_url
+        self.organization = organization
+        self.prefix = prefix
+        self.suffix = suffix
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            api_key=self.api_key,
```

Comment:
I copied it from [`GPT35Generator`](https://github.com/deepset-ai/haystack/blob/6dd52d91b2c0d8db9682b8d2901be99b098c8d42/haystack/preview/components/generators/openai/gpt35.py#L107-L109), but I see what you mean...

It should be addressed systematically for all components that have API keys.

Reply:
I suggested a better strategy some time ago [here](https://github.com/deepset-ai/haystack/pull/5712#discussion_r1315713514).

Reply:
@silvanocerza your idea seems good to me, but the example is incomplete.

Are you suggesting something similar to the following?
```python
    def to_dict(self):
        os.environ["COMPONENT_SECRET"] = "real-secret-value"
        return {"my_secret": "COMPONENT_SECRET", "other_field": self.other_field}
    
    @classmethod
    def from_dict(cls, data):
        data["my_secret"] = os.environ[data["my_secret"]]
        return cls(data)
```

Reply:
No. Don't update the environment when serialising, it's not necessary.

If you're serialising the component it doesn't necessarily mean that you're going to deserialise it back in the same environment, or even with the same env var.

Also this pollutes the user env in an hidden way. I wouldn't expect this side effect from a `to_dict()` method.

This is dangerous in several way as it could cause leaking secrets so don't do it.

Reply:
Ok. I understand.

So, for example, for a component that requires the OpenAI API key, should it be similar to this?
```python
    def to_dict(self):
        return {"my_secret": "OPENAI_API_KEY", "other_field": self.other_field}
    
    @classmethod
    def from_dict(cls, data):
        data["my_secret"] = os.environ[data["my_secret"]]
        return cls(data)
```

If the user does not have that ENV variable at the time of deserialization, do we let the serialization fail?

Reply:
Yes, deserialization should fail with a clear message letting the user know that this Component needs this env var.

Reply:
Not sure we're there yet. I see two use cases:
1) I use the env var, I never passed the key to the constructor. This is easy peasy, `self.api_key` should not be assigned.
2) I passed the key to the constructor and it gets assigned to `self.api_key`. I might or might not have an env var set

I would be surprised if I go with 2), serialize, then I find `self.api_key` was assigned to the env var upon deserialization. I think we should explicitly refuse to serialize the api_key if it was passed to the constructor, maybe suggesting to use the env var as a workaround.

Reply:
I tried to follow @silvanocerza's suggestion. (I also added several related tests).

Let me know if it can work...